File: android/app/src/main/java/com/zwj/timer/core/History.java
package com.zwj.timer.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/** 正计时的历史记录：最多保留最近 10 条，最新的在最前面。 */
public class History {
  public static final int DEFAULT_LIMIT = 10;

  public static class Entry {
    public String id = "";
    public long durationMs;
    public long endedAt;

    public String serialize() {
      List<String> fields = new ArrayList<>();
      fields.add(id);
      fields.add(Long.toString(durationMs));
      fields.add(Long.toString(endedAt));
      return LineStore.join(fields);
    }

    public static Entry deserialize(String line) {
      List<String> fields = LineStore.split(line);
      if (fields.size() < 3) return null;
      Entry entry = new Entry();
      entry.id = fields.get(0);
      try {
        entry.durationMs = Long.parseLong(fields.get(1));
        entry.endedAt = Long.parseLong(fields.get(2));
      } catch (NumberFormatException error) {
        return null;
      }
      if (entry.id.isEmpty() || entry.endedAt <= 0) return null;
      return entry;
    }
  }

  private final int limit;
  private final List<Entry> entries = new ArrayList<>();
  private long seq;

  public History() {
    this(DEFAULT_LIMIT);
  }

  public History(int limit) {
    this.limit = Math.max(0, limit);
  }

  public Entry add(long durationMs, long endedAt) {
    Entry entry = new Entry();
    seq++;
    entry.id = Long.toString(endedAt, 36) + "-" + Long.toString(seq, 36);
    entry.durationMs = Math.max(0, durationMs);
    entry.endedAt = endedAt;
    entries.add(0, entry);
    while (entries.size() > limit) entries.remove(entries.size() - 1);
    return entry;
  }

  public boolean remove(String id) {
    for (int i = 0; i < entries.size(); i++) {
      if (entries.get(i).id.equals(id)) {
        entries.remove(i);
        return true;
      }
    }
    return false;
  }

  public void clear() {
    entries.clear();
  }

  public List<Entry> list() {
    return new ArrayList<>(entries);
  }

  public int size() {
    return entries.size();
  }

  public int limit() {
    return limit;
  }

  public void loadFrom(List<String> lines) {
    entries.clear();
    for (String line : lines) {
      Entry entry = Entry.deserialize(line);
      if (entry == null) continue;
      entries.add(entry);
    }
    Collections.sort(
        entries,
        new Comparator<Entry>() {
          @Override
          public int compare(Entry a, Entry b) {
            return Long.compare(b.endedAt, a.endedAt);
          }
        });
    while (entries.size() > limit) entries.remove(entries.size() - 1);
  }

  public List<String> serializeLines() {
    List<String> lines = new ArrayList<>();
    for (Entry entry : entries) lines.add(entry.serialize());
    return lines;
  }
}

File: src/history.js
export const STORAGE_KEY = 'timer.history.v1';
export const DEFAULT_LIMIT = 10;

/** 仅用于测试或存储不可用时的兜底。 */
export function createMemoryStorage(initial = {}) {
  const map = new Map(Object.entries(initial));
  return {
    getItem: (key) => (map.has(key) ? map.get(key) : null),
    setItem: (key, value) => {
      map.set(key, String(value));
    },
    removeItem: (key) => {
      map.delete(key);
    },
  };
}

function isEntry(value) {
  return Boolean(value) &&
    typeof value === 'object' &&
    typeof value.id === 'string' &&
    value.id.length > 0 &&
    Number.isFinite(value.durationMs) &&
    value.durationMs >= 0 &&
    Number.isFinite(value.endedAt) &&
    value.endedAt > 0;
}

/**
 * 清洗外来数据：非法条目剔除、重复 id 剔除、按结束时刻倒序、截断到上限。
 */
export function sanitizeHistory(raw, limit = DEFAULT_LIMIT) {
  if (!Array.isArray(raw)) return [];
  const seen = new Set();
  const entries = [];
  for (const item of raw) {
    if (!isEntry(item) || seen.has(item.id)) continue;
    seen.add(item.id);
    entries.push({
      id: item.id,
      durationMs: Math.floor(item.durationMs),
      endedAt: item.endedAt,
    });
  }
  entries.sort((a, b) => b.endedAt - a.endedAt);
  return entries.slice(0, Math.max(0, limit));
}

export function createHistory({
  storage,
  limit = DEFAULT_LIMIT,
  now = Date.now,
  onStorageError,
} = {}) {
  const store =
    storage ?? (typeof globalThis.localStorage !== 'undefined' ? globalThis.localStorage : createMemoryStorage());

  let errorReported = false;
  let seq = 0;
  let entries = [];

  function reportError(error) {
    if (errorReported) return;
    errorReported = true;
    if (typeof onStorageError === 'function') onStorageError(error);
  }

  function readNow() {
    const value = now();
    return Number.isFinite(value) ? value : Date.now();
  }

  function load() {
    let raw = null;
    try {
      raw = store.getItem(STORAGE_KEY);
    } catch (error) {
      reportError(error);
      return [];
    }
    if (!raw) return [];
    try {
      return sanitizeHistory(JSON.parse(raw), limit);
    } catch (error) {
      reportError(error);
      try {
        store.removeItem(STORAGE_KEY);
      } catch {
        /* 清理失败不影响使用 */
      }
      return [];
    }
  }

  function persist() {
    try {
      store.setItem(STORAGE_KEY, JSON.stringify(entries));
    } catch (error) {
      reportError(error);
    }
  }

  function nextId(endedAt) {
    seq += 1;
    let suffix = seq;
    let id = `${endedAt.toString(36)}-${suffix.toString(36)}`;
    while (entries.some((entry) => entry.id === id)) {
      suffix += 1;
      id = `${endedAt.toString(36)}-${suffix.toString(36)}`;
    }
    return id;
  }

  function add({ durationMs, endedAt } = {}) {
    const stamp = Number.isFinite(endedAt) ? endedAt : readNow();
    const entry = {
      id: nextId(stamp),
      durationMs: Number.isFinite(durationMs) && durationMs > 0 ? Math.floor(durationMs) : 0,
      endedAt: stamp,
    };
    entries = [entry, ...entries].slice(0, Math.max(0, limit));
    persist();
    return { ...entry };
  }

  function remove(id) {
    const next = entries.filter((entry) => entry.id !== id);
    if (next.length === entries.length) return false;
    entries = next;
    persist();
    return true;
  }

  function clear() {
    entries = [];
    persist();
  }

  function reload() {
    entries = load();
    return list();
  }

  function list() {
    return entries.map((entry) => ({ ...entry }));
  }

  entries = load();

  return {
    list,
    add,
    remove,
    clear,
    reload,
    get limit() {
      return limit;
    },
  };
}
